## Segmenting and Clustering Neighborhoods in Toronto

### Coursera Capstone Project - Ramón De L'Hotellerie

#### Assignment 1 - Wikipedia html table into a DataFrame

In [18]:
from bs4 import BeautifulSoup as bsoup
from urllib.request import urlopen as uReq
import requests
import lxml
import pandas as pd
from pandas import DataFrame
import numpy as np

In [19]:
my_url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [20]:
r=requests.get(my_url)

#### Parsing html file with BeautifulSoup package

In [21]:
# Parse the htlm with Soup
page=bsoup(r.text,"html.parser")
page

<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className=document.documentElement.className.replace(/(^|\s)client-nojs(\s|$)/,"$1client-js$2");RLCONF={"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":900271985,"wgRevisionId":900271985,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Communications in Ontario","Postal codes in Canada","Toronto","Ontario-related lists"],"wgBreakFrames":!1,"wgPageContentLanguage":"en","wgPageContentModel":"wikitext","wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","

#### Finding the table

In [22]:
rtable=page.table
rtable

<table class="wikitable sortable">
<tbody><tr>
<th>Postcode</th>
<th>Borough</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
</td></tr>
<tr>
<td>M4A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Harbourfront_(Toronto)" title="Harbourfront (Toronto)">Harbourfront</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Regent_Park" title="Regent Park">Regent Park</a>
</td></tr>
<tr>
<td>M6A</td>

In [23]:
totals=rtable.find_all('tr')
nrows=len(totals)
nrows

289

#### The totals of rows of data is 289

In [27]:
totals[0:10]

[<tr>
 <th>Postcode</th>
 <th>Borough</th>
 <th>Neighbourhood
 </th></tr>, <tr>
 <td>M1A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M2A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M3A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
 </td></tr>, <tr>
 <td>M4A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
 </td></tr>, <tr>
 <td>M5A</td>
 <td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
 <td><a href="/wiki/Harbourfront_(Toronto)" title="Harbourfront (Toronto)">Harbourfront</a>
 </td></tr>, <tr>
 <td>M5A</td>
 <td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
 <td><a href="/wiki/Regent_Park" title="Regent Park">Regent Park</a>
 </td></tr>, <tr>
 <td>M6A</td>
 <td

#### The Header of the DataFrame is the first row of data

In [28]:
header=totals[0].text.split()
header

['Postcode', 'Borough', 'Neighbourhood']

#### Checking row 27 as sample 

In [29]:
totals[27].text

'\nM9B\nEtobicoke\nWest Deane Park\n'

In [30]:
totals[27].text.split('\n')

['', 'M9B', 'Etobicoke', 'West Deane Park', '']

In [31]:
Postcode=totals[27].text.split('\n')[1]
Postcode

'M9B'

In [32]:
Borough=totals[27].text.split('\n')[2]
Borough

'Etobicoke'

In [33]:
Neighborhood=totals[27].text.split('\n')[3]
Neighborhood

'West Deane Park'

#### Extracting cells into a dataframe

In [34]:
# iteration loop to harvest all records

records =[]
n=1
while n < nrows :
    Postcode=totals[n].text.split('\n')[1]
    Borough=totals[n].text.split('\n')[2]
    Neighborhood=totals[n].text.split('\n')[3]
    records.append((Postcode, Borough,Neighborhood))
    n=n+1

df=pd.DataFrame(records, columns=['PostalCode', 'Borough', 'Neighbourhood'])
df.head(5)

PostalCode           Borough     Neighbourhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

In [35]:
df.shape

(288, 3)

##### So the dataframe has exactly the same number of rows as the original table

#### Removing rows with Borough = 'Not assigned'

In [36]:
# Rows with Borough equaling 'Not assigned'
df[df['Borough']=='Not assigned'].count()

PostalCode       77
Borough          77
Neighbourhood    77
dtype: int64

In [37]:
# drop rows where 'Not assigned' appears in column '[Borough]'
df1=df[~df.Borough.str.contains("Not assigned")]
df1=df1.reset_index(drop=True)

In [38]:
df1.head(15)

PostalCode           Borough     Neighbourhood
0         M3A        North York         Parkwoods
1         M4A        North York  Victoria Village
2         M5A  Downtown Toronto      Harbourfront
3         M5A  Downtown Toronto       Regent Park
4         M6A        North York  Lawrence Heights
5         M6A        North York    Lawrence Manor
6         M7A      Queen's Park      Not assigned
7         M9A         Etobicoke  Islington Avenue
8         M1B       Scarborough             Rouge
9         M1B       Scarborough           Malvern
10        M3B        North York   Don Mills North
11        M4B         East York  Woodbine Gardens
12        M4B         East York     Parkview Hill
13        M5B  Downtown Toronto           Ryerson
14        M5B  Downtown Toronto   Garden District

In [39]:
df1.tail(15)

PostalCode           Borough  \
196        M4Y  Downtown Toronto   
197        M7Y      East Toronto   
198        M8Y         Etobicoke   
199        M8Y         Etobicoke   
200        M8Y         Etobicoke   
201        M8Y         Etobicoke   
202        M8Y         Etobicoke   
203        M8Y         Etobicoke   
204        M8Y         Etobicoke   
205        M8Y         Etobicoke   
206        M8Z         Etobicoke   
207        M8Z         Etobicoke   
208        M8Z         Etobicoke   
209        M8Z         Etobicoke   
210        M8Z         Etobicoke   

                                         Neighbourhood  
196                               Church and Wellesley  
197  Business Reply Mail Processing Centre 969 Eastern  
198                                         Humber Bay  
199                                   King's Mill Park  
200                           Kingsway Park South East  
201                                          Mimico NE  
202                                     Old Mill South  
203                                 The Queensway East  
204                              Royal York South East  
205                                           Sunnylea  
206                           Kingsway Park South West  
207                                          Mimico NW  
208                                 The Queensway West  
209                              Royal York South West  
210                                     South of Bloor

In [40]:
df1.shape

(211, 3)

##### So there are no rows with Not Assigned data under the column Borough, and the remaining number of rows is now 211

#### Replacing 'Not assigned' neighborhoods with the name of the Borough

In [41]:
df1.loc[df1['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df1['Borough']

In [42]:
df1.head(15) 

PostalCode           Borough     Neighbourhood
0         M3A        North York         Parkwoods
1         M4A        North York  Victoria Village
2         M5A  Downtown Toronto      Harbourfront
3         M5A  Downtown Toronto       Regent Park
4         M6A        North York  Lawrence Heights
5         M6A        North York    Lawrence Manor
6         M7A      Queen's Park      Queen's Park
7         M9A         Etobicoke  Islington Avenue
8         M1B       Scarborough             Rouge
9         M1B       Scarborough           Malvern
10        M3B        North York   Don Mills North
11        M4B         East York  Woodbine Gardens
12        M4B         East York     Parkview Hill
13        M5B  Downtown Toronto           Ryerson
14        M5B  Downtown Toronto   Garden District

In [43]:
df1.shape

(211, 3)

##### So there are no rows with Not Assigned data under any of the columns

#### Consolidating dataframe to unique Postal Codes

In [44]:
postalcodes = df1['PostalCode'].nunique()
boroughs = df1['Borough'].nunique()
neighbourhoods= df1['Neighbourhood'].nunique()
print('Same Postalcodes : ' + str(postalcodes))
print('Same Boroughs  : '+ str(boroughs))
print('Same Neighbourhoods  :' + str(neighbourhoods))

Same Postalcodes : 103
Same Boroughs  : 11
Same Neighbourhoods  :209


##### There are 103 Same Postcodes and 11 Borough with 209 Neighbourhoods with the same data

In [46]:
df1.groupby(['PostalCode','Borough','Neighbourhood']).size().reset_index(name='Count').head()

PostalCode      Borough   Neighbourhood  Count
0        M1B  Scarborough         Malvern      1
1        M1B  Scarborough           Rouge      1
2        M1C  Scarborough  Highland Creek      1
3        M1C  Scarborough      Port Union      1
4        M1C  Scarborough      Rouge Hill      1

#### Consolidating the dataframe to same PostalCodes and Neighbourhoods

In [47]:
df1.head(15)

PostalCode           Borough     Neighbourhood
0         M3A        North York         Parkwoods
1         M4A        North York  Victoria Village
2         M5A  Downtown Toronto      Harbourfront
3         M5A  Downtown Toronto       Regent Park
4         M6A        North York  Lawrence Heights
5         M6A        North York    Lawrence Manor
6         M7A      Queen's Park      Queen's Park
7         M9A         Etobicoke  Islington Avenue
8         M1B       Scarborough             Rouge
9         M1B       Scarborough           Malvern
10        M3B        North York   Don Mills North
11        M4B         East York  Woodbine Gardens
12        M4B         East York     Parkview Hill
13        M5B  Downtown Toronto           Ryerson
14        M5B  Downtown Toronto   Garden District

In [48]:
nrows1=len(df1)
nrows1

211

#### Using a new dataframe and keeping safe df1

In [49]:
df2=df1
df2.head(15)

PostalCode           Borough     Neighbourhood
0         M3A        North York         Parkwoods
1         M4A        North York  Victoria Village
2         M5A  Downtown Toronto      Harbourfront
3         M5A  Downtown Toronto       Regent Park
4         M6A        North York  Lawrence Heights
5         M6A        North York    Lawrence Manor
6         M7A      Queen's Park      Queen's Park
7         M9A         Etobicoke  Islington Avenue
8         M1B       Scarborough             Rouge
9         M1B       Scarborough           Malvern
10        M3B        North York   Don Mills North
11        M4B         East York  Woodbine Gardens
12        M4B         East York     Parkview Hill
13        M5B  Downtown Toronto           Ryerson
14        M5B  Downtown Toronto   Garden District

In [50]:
nrows2=len(df2)-1
nrows2

210

##### So the number of rows of this second dataframe is also 211

#### Extracting consolidated dataframe df2 with same postalcodes

In [51]:
n=0

while n < nrows2 :
    post1=df2.iloc[n,0]
    #post1
    m=n+1
    post2=df2.iloc[m,0]
    #post2
    neigh1=df2.iloc[n,2]
    neigh2=df2.iloc[m,2]
    if post1==post2:
        df2.Neighbourhood[n,2] = neigh1=neigh1+','+neigh2
        #df2 = df2[df2.Neighbourhood != 'neigh2']
        df2=df2.drop(df2.index[m])
        nrows2=nrows2-1
        df2 = df2.reset_index(drop=True)
    else:
        n=n+1


#df2 = df2.reset_index(drop=True)
# print(post1, post2, 'Nigh1 is...'+ neigh1,',,,,,,,','Nigh2 is...' + neigh2, n,m)
df2.index

RangeIndex(start=0, stop=103, step=1)

In [52]:
df2.head(15)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A      Queen's Park   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   

                                        Neighbourhood  
0                                           Parkwoods  
1                                    Victoria Village  
2   Kingsway Park South West,Mimico NW,The Queensw...  
3                     Lawrence Heights,Lawrence Manor  
4                                        Queen's Park  
5                                    Islington Avenue  
6                                       Rouge,Malvern  
7                                     Don Mills North  
8                      Woodbine Gardens,Parkview Hill  
9                             Ryerson,Garden District  
10                                          Glencairn  
11  Cloverdale,Islington,Martin Grove,Princess Gar...  
12               Highland Creek,Rouge Hill,Port Union  
13                    Flemingdon Park,Don Mills South  
14                                   Woodbine Heights

In [53]:
df2.index

RangeIndex(start=0, stop=103, step=1)

In [54]:
df2.shape

(103, 3)

##### So df2 has 103 rows and 3 columns

### <><><><><><><><><>     <><><><><><><><><>     <><><><><><><><><>   

#### Assignment 2 - Calculating Latitude and Longitude for PostalCodes 

##### The geocoder did not work, so I used the cvs file

In [55]:
# Read the cvs file and convert it to a dataframe
#df_pcodes=pd.read_csv('Toronto_Geospatial_Coordinates.csv')
#df_pcodes.columns = ['PostalCode','Latitude','Longitude']
#df_pcodes.head()

url='http://cocl.us/Geospatial_data'
df_pcodes=pd.read_csv(url)
df_pcodes.head(15)

Postal Code   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
5          M1J  43.744734 -79.239476
6          M1K  43.727929 -79.262029
7          M1L  43.711112 -79.284577
8          M1M  43.716316 -79.239476
9          M1N  43.692657 -79.264848
10         M1P  43.757410 -79.273304
11         M1R  43.750072 -79.295849
12         M1S  43.794200 -79.262029
13         M1T  43.781638 -79.304302
14         M1V  43.815252 -79.284577

##### Converting column "Postal Code" on to "PostalCode"

In [56]:
df_pcodes.columns = ['PostalCode', 'Latitude', 'Longitude']
df_pcodes.head(15)

PostalCode   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476
5         M1J  43.744734 -79.239476
6         M1K  43.727929 -79.262029
7         M1L  43.711112 -79.284577
8         M1M  43.716316 -79.239476
9         M1N  43.692657 -79.264848
10        M1P  43.757410 -79.273304
11        M1R  43.750072 -79.295849
12        M1S  43.794200 -79.262029
13        M1T  43.781638 -79.304302
14        M1V  43.815252 -79.284577

##### Sorting df2 per PostalCode and merging it with the df_pcodes 

In [57]:
df2s=df2.sort_values('PostalCode')

In [58]:
neighborhoods=pd.merge(df2s,df_pcodes, how='right', on = 'PostalCode')
neighborhoods.head(15)

PostalCode      Borough                                      Neighbourhood  \
0         M1B  Scarborough                                      Rouge,Malvern   
1         M1C  Scarborough               Highland Creek,Rouge Hill,Port Union   
2         M1E  Scarborough                    Guildwood,Morningside,West Hill   
3         M1G  Scarborough                                             Woburn   
4         M1H  Scarborough                                          Cedarbrae   
5         M1J  Scarborough                                Scarborough Village   
6         M1K  Scarborough          East Birchmount Park,Ionview,Kennedy Park   
7         M1L  Scarborough                      Clairlea,Golden Mile,Oakridge   
8         M1M  Scarborough      Cliffcrest,Cliffside,Scarborough Village West   
9         M1N  Scarborough                         Birch Cliff,Cliffside West   
10        M1P  Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...   
11        M1R  Scarborough                                   Maryvale,Wexford   
12        M1S  Scarborough                                          Agincourt   
13        M1T  Scarborough              Clarks Corners,Sullivan,Tam O'Shanter   
14        M1V  Scarborough  Agincourt North,L'Amoreaux East,Milliken,Steel...   

     Latitude  Longitude  
0   43.806686 -79.194353  
1   43.784535 -79.160497  
2   43.763573 -79.188711  
3   43.770992 -79.216917  
4   43.773136 -79.239476  
5   43.744734 -79.239476  
6   43.727929 -79.262029  
7   43.711112 -79.284577  
8   43.716316 -79.239476  
9   43.692657 -79.264848  
10  43.757410 -79.273304  
11  43.750072 -79.295849  
12  43.794200 -79.262029  
13  43.781638 -79.304302  
14  43.815252 -79.284577

In [59]:
neighborhoods.shape

(103, 5)

##### So there are 103 rows after merging the dataframe with the cvs, with 5 columns after adding Latitude and Longitude

### <><><><><><><><><>     <><><><><><><><><>     <><><><><><><><><>  

#### Assignment 3 - Clustering Toronto's neighborhoods

In [60]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: | 
The environment is inconsistent, please check the package plan carefully
The following packages are causing the inconsistency:

  - defaults/linux-64::anaconda==5.3.1=py37_0
  - defaults/linux-64::astropy==3.0.4=py37h14c3975_0
  - defaults/linux-64::bkcharts==0.2=py37_0
  - defaults/linux-64::blaze==0.11.3=py37_0
  - defaults/linux-64::bokeh==0.13.0=py37_0
  - defaults/linux-64::bottleneck==1.2.1=py37h035aef0_1
  - defaults/linux-64::dask==0.19.1=py37_0
  - defaults/linux-64::datashape==0.5.4=py37_1
  - defaults/linux-64::mkl-service==1.1.2=py37h90e4bf4_5
  - defaults/linux-64::numba==0.39.0=py37h04863e7_0
  - defaults/linux-64::numexpr==2.6.8=py37hd89afb7_0
  - defaults/linux-64::odo==0.5.1=py37_0
  - defaults/linux-64::pytables==3.4.4=py37ha205bf6_0
  - defaults/linux-64::pytest-arraydiff==0.2=py37h39e3cac_0
  - defaults/linux-64::pytest-astropy==0.4.0=py37_0
  - defaults/linux-64::pytest-doctestplus==0.1.3=py37_0
  - defaults/linux-64::pywavelets==1.0.0=py37h

##### Selecting rows where Borough includes the word Toronto

In [61]:
toronto_totals= neighborhoods[neighborhoods['Borough'].str.contains('Toronto', na = False)].reset_index(drop=True)
toronto_totals.head(15)

PostalCode           Borough  \
0         M4E      East Toronto   
1         M4K      East Toronto   
2         M4L      East Toronto   
3         M4M      East Toronto   
4         M4N   Central Toronto   
5         M4P   Central Toronto   
6         M4R   Central Toronto   
7         M4S   Central Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
10        M4W  Downtown Toronto   
11        M4X  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
14        M5B  Downtown Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                         The Danforth West,Riverdale  43.679557 -79.352188  
2                       The Beaches West,India Bazaar  43.668999 -79.315572  
3                                     Studio District  43.659526 -79.340923  
4                                       Lawrence Park  43.728020 -79.388790  
5                                    Davisville North  43.712751 -79.390197  
6                                  North Toronto West  43.715383 -79.405678  
7                                          Davisville  43.704324 -79.388790  
8                          Moore Park,Summerhill East  43.689574 -79.383160  
9   Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049  
10                                           Rosedale  43.679563 -79.377529  
11                         Cabbagetown,St. James Town  43.667967 -79.367675  
12                               Church and Wellesley  43.665860 -79.383160  
13  Kingsway Park South West,Mimico NW,The Queensw...  43.654260 -79.360636  
14                            Ryerson,Garden District  43.657162 -79.378937

In [62]:
toronto_totals.shape

(38, 5)

##### So there are 38 Boroughs including the word Toronto

#### Mapping Toronto

In [63]:
latitude = 43.6532
longitude= -79.3832

In [71]:
# create map of TORONTO using latitude and longitude values above:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_totals['Latitude'], toronto_totals['Longitude'], toronto_totals['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Foursquare API to explore and segment

##### My Foursquare Credentials

In [106]:
CLIENT_ID = 'NY50URDZQUKHFXKFB2L2ETTX0P5XXGHKHDFN3Y5VFVAM4RH5' # your Foursquare ID
CLIENT_SECRET = 'HI5IGG4WXAF1E53B5DKLLXZ4PMHY55RWWZ0IK4PRFQVOLKNR' # your Foursquare Secret
VERSION = '20180604'
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: NY50URDZQUKHFXKFB2L2ETTX0P5XXGHKHDFN3Y5VFVAM4RH5
CLIENT_SECRET:HI5IGG4WXAF1E53B5DKLLXZ4PMHY55RWWZ0IK4PRFQVOLKNR


In [107]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

In [108]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

##### Writting the code to run the function on each neighborhood and create a new dataframe called toronto_venues.

In [109]:
toronto_venues = getNearbyVenues(names=toronto_totals['Neighbourhood'],
                                   latitudes=toronto_totals['Latitude'],
                                   longitudes=toronto_totals['Longitude']
                                    )

The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Kingsway Park South West,Mimico NW,The Queensway West,Royal York South West,South of Bloor
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place

In [111]:
print(toronto_venues.shape)
toronto_venues.head(15)

(1697, 7)


Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4   The Danforth West,Riverdale              43.679557   
5   The Danforth West,Riverdale              43.679557   
6   The Danforth West,Riverdale              43.679557   
7   The Danforth West,Riverdale              43.679557   
8   The Danforth West,Riverdale              43.679557   
9   The Danforth West,Riverdale              43.679557   
10  The Danforth West,Riverdale              43.679557   
11  The Danforth West,Riverdale              43.679557   
12  The Danforth West,Riverdale              43.679557   
13  The Danforth West,Riverdale              43.679557   
14  The Danforth West,Riverdale              43.679557   

    Neighborhood Longitude                               Venue  \
0               -79.293031                   Glen Manor Ravine   
1               -79.293031  The Big Carrot Natural Food Market   
2               -79.293031                 Grover Pub and Grub   
3               -79.293031                       Upper Beaches   
4               -79.352188                            Pantheon   
5               -79.352188                        Dolce Gelato   
6               -79.352188                       MenEssentials   
7               -79.352188                     Cafe Fiorentina   
8               -79.352188                          La Diperie   
9               -79.352188                               Mezes   
10              -79.352188             Messini Authentic Gyros   
11              -79.352188              Louis Cifer Brew Works   
12              -79.352188                Moksha Yoga Danforth   
13              -79.352188         Christina's On The Danforth   
14              -79.352188                   The Auld Spot Pub   

    Venue Latitude  Venue Longitude      Venue Category  
0        43.676821       -79.293942               Trail  
1        43.678879       -79.297734   Health Food Store  
2        43.679181       -79.297215                 Pub  
3        43.680563       -79.292869        Neighborhood  
4        43.677621       -79.351434    Greek Restaurant  
5        43.677773       -79.351187      Ice Cream Shop  
6        43.677820       -79.351265      Cosmetics Shop  
7        43.677743       -79.350115  Italian Restaurant  
8        43.677530       -79.352295      Ice Cream Shop  
9        43.677962       -79.350196    Greek Restaurant  
10       43.677827       -79.350569    Greek Restaurant  
11       43.677663       -79.351313             Brewery  
12       43.677622       -79.352116         Yoga Studio  
13       43.678240       -79.349185    Greek Restaurant  
14       43.677335       -79.353130                 Pub

##### Number of venues per Neiborhood

In [112]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Berczy Park                                                            55   
Brockton,Exhibition Place,Parkdale Village                             21   
Business Reply Mail Processing Centre 969 Eastern                      19   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     15   
Cabbagetown,St. James Town                                             46   
Central Bay Street                                                     88   
Chinatown,Grange Park,Kensington Market                               100   
Christie                                                               15   
Church and Wellesley                                                   87   
Commerce Court,Victoria Hotel                                         100   
Davisville                                                             38   
Davisville North                                                        9   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     14   
Design Exchange,Toronto Dominion Centre                               100   
Dovercourt Village,Dufferin                                            20   
First Canadian Place,Underground city                                 100   
Forest Hill North,Forest Hill West                                      4   
Harbord,University of Toronto                                          34   
Harbourfront East,Toronto Islands,Union Station                       100   
High Park,The Junction South                                           23   
Kingsway Park South West,Mimico NW,The Queenswa...                     48   
Lawrence Park                                                           3   
Little Portugal,Trinity                                                62   
Moore Park,Summerhill East                                              2   
North Toronto West                                                     16   
Parkdale,Roncesvalles                                                  15   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede,Swansea                                                      38   
Ryerson,Garden District                                               100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        94   
Studio District                                                        38   
The Annex,North Midtown,Yorkville                                      23   
The Beaches                                                             4   
The Beaches West,India Bazaar                                          18   
The Danforth West,Riverdale                                            42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             55   
Brockton,Exhibition Place,Parkdale Village                              21   
Business Reply Mail Processing Centre 969 Eastern                       19   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      15   
Cabbagetown,St. James Town                                              46   
Central Bay Street                                                      88   
Chinatown,Grange Park,Kensington Market                                100   
Christie                                                                15   
Church and Wellesley                                                    87   
Commerce Court,Victoria Hotel      

##### Number of unique categories

In [113]:
print('The number of unique categories is {}.'.format(len(toronto_venues['Venue Category'].unique())))

The number of unique categories is 237.


#### Analyzing Neighborhoods

In [115]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
cols=list(toronto_onehot.columns.values)
cols.pop(cols.index('Neighborhood'))
toronto_onehot=toronto_onehot[['Neighborhood']+cols]

# rename Neighborhood for Neighbourhood so that future merge works
toronto_onehot.rename(columns = {'Neighborhood': 'Neighbourhood'}, inplace = True)
toronto_onehot.head(15)

Neighbourhood  Adult Boutique  Afghan Restaurant  Airport  \
0                   The Beaches               0                  0        0   
1                   The Beaches               0                  0        0   
2                   The Beaches               0                  0        0   
3                   The Beaches               0                  0        0   
4   The Danforth West,Riverdale               0                  0        0   
5   The Danforth West,Riverdale               0                  0        0   
6   The Danforth West,Riverdale               0                  0        0   
7   The Danforth West,Riverdale               0                  0        0   
8   The Danforth West,Riverdale               0                  0        0   
9   The Danforth West,Riverdale               0                  0        0   
10  The Danforth West,Riverdale               0                  0        0   
11  The Danforth West,Riverdale               0                  0        0   
12  The Danforth West,Riverdale               0                  0        0   
13  The Danforth West,Riverdale               0                  0        0   
14  The Danforth West,Riverdale               0                  0        0   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                    0             0               0                0   
1                    0             0               0                0   
2                    0             0               0                0   
3                    0             0               0                0   
4                    0             0               0                0   
5                    0             0               0                0   
6                    0             0               0                0   
7                    0             0               0                0   
8                    0             0               0                0   
9                    0             0               0                0   
10                   0             0               0                0   
11                   0             0               0                0   
12                   0             0               0                0   
13                   0             0               0                0   
14                   0             0               0                0   

    Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                  0                    0             0         0   
1                  0                    0             0         0   
2                  0                    0             0         0   
3                  0                    0             0         0   
4                  0                    0             0         0   
5                  0                    0             0         0   
6                  0                    0             0         0   
7                  0                    0             0         0   
8                  0                    0             0         0   
9                  0                    0             0         0   
10                 0                    0             0         0   
11                 0                    0             0         0   
12                 0                    0             0         0   
13                 0                    0             0         0   
14                 0                    0             0         0   

    Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0             0           0                    0                 0   
1             0           0                    0                 0   
2             0           0                    0                 0   
3             0           0                    0                 0   
4             0           0                    0                 0   
5             0           0                    0           

In [116]:
toronto_onehot.shape

(1697, 237)

#### Grouping rows by neighborhood and by the mean of the frequency of occurrence of each category

In [117]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Adult Boutique  \
0                              Adelaide,King,Richmond        0.000000   
1                                         Berczy Park        0.000000   
2          Brockton,Exhibition Place,Parkdale Village        0.000000   
3   Business Reply Mail Processing Centre 969 Eastern        0.000000   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...        0.000000   
5                          Cabbagetown,St. James Town        0.000000   
6                                  Central Bay Street        0.000000   
7             Chinatown,Grange Park,Kensington Market        0.000000   
8                                            Christie        0.000000   
9                                Church and Wellesley        0.011494   
10                      Commerce Court,Victoria Hotel        0.000000   
11                                         Davisville        0.000000   
12                                   Davisville North        0.000000   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...        0.000000   
14            Design Exchange,Toronto Dominion Centre        0.000000   
15                        Dovercourt Village,Dufferin        0.000000   
16              First Canadian Place,Underground city        0.000000   
17                 Forest Hill North,Forest Hill West        0.000000   
18                      Harbord,University of Toronto        0.000000   
19    Harbourfront East,Toronto Islands,Union Station        0.000000   
20                       High Park,The Junction South        0.000000   
21  Kingsway Park South West,Mimico NW,The Queensw...        0.000000   
22                                      Lawrence Park        0.000000   
23                            Little Portugal,Trinity        0.000000   
24                         Moore Park,Summerhill East        0.000000   
25                                 North Toronto West        0.000000   
26                              Parkdale,Roncesvalles        0.000000   
27                                           Rosedale        0.000000   
28                                           Roselawn        0.000000   
29                                  Runnymede,Swansea        0.000000   
30                            Ryerson,Garden District        0.000000   
31                                     St. James Town        0.000000   
32                    Stn A PO Boxes 25 The Esplanade        0.000000   
33                                    Studio District        0.000000   
34                  The Annex,North Midtown,Yorkville        0.000000   
35                                        The Beaches        0.000000   
36                      The Beaches West,India Bazaar        0.000000   
37                        The Danforth West,Riverdale        0.000000   

    Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0            0.000000  0.000000            0.000000      0.000000   
1            0.000000  0.000000            0.000000      0.000000   
2            0.000000  0.000000            0.000000      0.000000   
3            0.000000  0.000000            0.000000      0.000000   
4            0.000000  0.066667            0.066667      0.066667   
5            0.000000  0.000000            0.000000      0.000000   
6            0.000000  0.000000            0.000000      0.000000   
7            0.000000  0.000000            0.000000      0.000000   
8            0.000000  0.000000            0.000000      0.000000   
9            0.011494  0.000000            0.000000      0.000000   
10           0.000000  0.000000            0.000000      0.000000   
11           0.000000  0.000000            0.000000      0.000000   
12           0.000000  0.000000            0.000000      0.000000   
13           0.000000  0.000000            0.000000      0.000000   
14           0.000000  0.000000            0.000000      0.000000   
15           0.000000  0.000000            0.000000      0.000000   
16           0.0000

In [118]:
toronto_grouped.shape

(38, 237)

#### Printing neighborhoods along with the top 5 most common venues

In [119]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
                 venue  freq
0                 Café  0.05
1          Coffee Shop  0.05
2                  Bar  0.04
3           Steakhouse  0.04
4  American Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2              Bakery  0.04
3         Cheese Shop  0.04
4  Seafood Restaurant  0.04


----Brockton,Exhibition Place,Parkdale Village----
            venue  freq
0     Coffee Shop  0.10
1  Breakfast Spot  0.10
2            Café  0.10
3             Gym  0.05
4         Stadium  0.05


----Business Reply Mail Processing Centre 969 Eastern----
           venue  freq
0    Yoga Studio  0.05
1     Restaurant  0.05
2            Spa  0.05
3  Burrito Place  0.05
4        Butcher  0.05


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0    Airport Lounge  0.13
1   Airport Service  0.13
2  Airport Terminal  

In [124]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [125]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighbourhood 1st Most Common Venue  \
0                              Adelaide,King,Richmond           Coffee Shop   
1                                         Berczy Park           Coffee Shop   
2          Brockton,Exhibition Place,Parkdale Village           Coffee Shop   
3   Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...      Airport Terminal   
5                          Cabbagetown,St. James Town           Coffee Shop   
6                                  Central Bay Street           Coffee Shop   
7             Chinatown,Grange Park,Kensington Market                  Café   
8                                            Christie         Grocery Store   
9                                Church and Wellesley           Coffee Shop   
10                      Commerce Court,Victoria Hotel           Coffee Shop   
11                                         Davisville           Pizza Place   
12                                   Davisville North            Playground   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...           Coffee Shop   
14            Design Exchange,Toronto Dominion Centre           Coffee Shop   
15                        Dovercourt Village,Dufferin              Pharmacy   
16              First Canadian Place,Underground city           Coffee Shop   
17                 Forest Hill North,Forest Hill West         Jewelry Store   
18                      Harbord,University of Toronto                  Café   
19    Harbourfront East,Toronto Islands,Union Station           Coffee Shop   
20                       High Park,The Junction South                   Bar   
21  Kingsway Park South West,Mimico NW,The Queensw...           Coffee Shop   
22                                      Lawrence Park                  Park   
23                            Little Portugal,Trinity                   Bar   
24                         Moore Park,Summerhill East            Playground   
25                                 North Toronto West           Coffee Shop   
26                              Parkdale,Roncesvalles        Breakfast Spot   
27                                           Rosedale                  Park   
28                                           Roselawn        Ice Cream Shop   
29                                  Runnymede,Swansea           Pizza Place   
30                            Ryerson,Garden District           Coffee Shop   
31                                     St. James Town           Coffee Shop   
32                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
33                                    Studio District                  Café   
34                  The Annex,North Midtown,Yorkville           Coffee Shop   
35                                        The Beaches     Health Food Store   
36                      The Beaches West,India Bazaar                  Park   
37                        The Danforth West,Riverdale      Greek Restaurant   

            2nd Most Common Venue 3rd Most Common Venue  \
0                            Café                   Bar   
1                    Cocktail Bar            Steakhouse   
2                            Café        Breakfast Spot   
3                     Pizza Place               Brewery   
4                  Airport Lounge       Airport Service   
5                      Restaurant           Pizza Place   
6              Italian Restaurant                  Café   
7   Vegetarian / Vegan Restaurant    Mexican Restaurant   
8                            Café                  Park   
9             Japanese Restaurant      Sushi Restaurant   
10                           Café                 Hotel   
11                   Dessert Shop        Sandwich Place   
12                          Hotel        Clothing Store   
13                            Pub   Fried Chicken Joint   
14                          Hotel                  Café   
15            

#### Clustering neighborhoods

##### New dataframe for clusters and top 10 venues for each neighborhood

In [128]:
toronto_merged = toronto_totals

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

PostalCode          Borough                  Neighbourhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0     Health Food Store                   Pub   
1 -79.352188               0      Greek Restaurant           Coffee Shop   
2 -79.315572               0                  Park  Fast Food Restaurant   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.388790               0                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0                 Trail        Discount Store     Filipino Restaurant   
1    Italian Restaurant        Ice Cream Shop  Furniture / Home Store   
2    Italian Restaurant           Pizza Place                     Gym   
3             Gastropub    Italian Restaurant                  Bakery   
4              Bus Line           Yoga Studio        Doner Restaurant   

  6th Most Common Venue 7th Most Common Venue    8th Most Common Venue  \
0  Fast Food Restaurant        Farmers Market       Falafel Restaurant   
1               Brewery                Bakery  Fruit & Vegetable Store   
2                   Pub         Movie Theater           Sandwich Place   
3   American Restaurant           Yoga Studio                  Brewery   
4   Filipino Restaurant  Fast Food Restaurant           Farmers Market   

  9th Most Common Venue 10th Most Common Venue  
0           Event Space   Ethiopian Restaurant  
1             Juice Bar           Liquor Store  
2         Burrito Place           Burger Joint  
3    Seafood Restaurant         Sandwich Place  
4    Falafel Restaurant            Event Space

#### Clustering Neighbourhoods

##### Running k-means to cluster the neighbourhood into 5 clusters

In [129]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

#### Clusters visualization of Toronto's Neighborhoods

In [131]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

##### Now, one can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, one can then assign a name to each cluster, if so desired

#### Cluster 1 - Most common  venues: Coffee shops, restaurants, pizza places and bookstores

In [132]:

toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               0     Health Food Store   
1       East Toronto               0      Greek Restaurant   
2       East Toronto               0                  Park   
3       East Toronto               0                  Café   
4    Central Toronto               0                  Park   
5    Central Toronto               0            Playground   
6    Central Toronto               0           Coffee Shop   
7    Central Toronto               0           Pizza Place   
8    Central Toronto               0            Playground   
9    Central Toronto               0           Coffee Shop   
10  Downtown Toronto               0                  Park   
11  Downtown Toronto               0           Coffee Shop   
12  Downtown Toronto               0           Coffee Shop   
13  Downtown Toronto               0           Coffee Shop   
14  Downtown Toronto               0           Coffee Shop   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0           Coffee Shop   
19  Downtown Toronto               0           Coffee Shop   
20  Downtown Toronto               0           Coffee Shop   
21  Downtown Toronto               0           Coffee Shop   
23   Central Toronto               0         Jewelry Store   
25  Downtown Toronto               0                  Café   
26  Downtown Toronto               0                  Café   
29  Downtown Toronto               0           Coffee Shop   
30  Downtown Toronto               0         Grocery Store   
31      West Toronto               0              Pharmacy   
32      West Toronto               0                   Bar   
33      West Toronto               0           Coffee Shop   
34      West Toronto               0                   Bar   
35      West Toronto               0        Breakfast Spot   
36      West Toronto               0           Pizza Place   
37      East Toronto               0           Yoga Studio   

            2nd Most Common Venue 3rd Most Common Venue  \
0                             Pub                 Trail   
1                     Coffee Shop    Italian Restaurant   
2            Fast Food Restaurant    Italian Restaurant   
3                     Coffee Shop             Gastropub   
4                     Swim School              Bus Line   
5                           Hotel        Clothing Store   
6                     Yoga Studio            Bagel Shop   
7                    Dessert Shop        Sandwich Place   
8                    Tennis Court          Concert Hall   
9                             Pub   Fried Chicken Joint   
10                     Playground                 Trail   
11                     Restaurant           Pizza Place   
12            Japanese Restaurant      Sushi Restaurant   
13                           Park                Bakery   
14                 Clothing Store        Cosmetics Shop   
15                           Café                 Hotel   
16                   Cocktail Bar            Steakhouse   
18                           Café                   Bar   
19                          Hotel              Aquarium   
20                          Hotel                  Café   
21                           Café                 Hotel   
23                          Trail      Sushi Restaurant   
25                      Bookstore                Bakery   
26  Vegetarian / Vegan Restaurant    Mexican Restaurant   
29                           Café                 Hotel   
30                           Café                  Park   
31                         Bakery           Supermarket   
32                    Coffee Shop      Asian Restaurant   
33                           Café        Breakfast Spot   
34             Mexican Restaurant                  Café   
35                      Gift Shop               Dog Run   
36              

#### Cluster 2 -  Most common venues: Coffee shops, Sandwich Place, Café and Pizza Place

In [133]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
24  Central Toronto               1           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
24        Sandwich Place                  Café           Pizza Place   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
24        History Museum     Indian Restaurant              Pharmacy   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
24                  Park                   Pub           Burger Joint

#### Cluster 3 - Most common venues: Coffee shop, Italian Restaurant, Café and Sandwich Place

In [134]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
17  Downtown Toronto               2           Coffee Shop   
27  Downtown Toronto               2      Airport Terminal   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17    Italian Restaurant                  Café        Sandwich Place   
27        Airport Lounge       Airport Service       Harbor / Marina   

   5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
17          Burger Joint  Middle Eastern Restaurant      Sushi Restaurant   
27      Sculpture Garden              Boat or Ferry              Boutique   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
17                Bakery    Chinese Restaurant                    Bar  
27                 Plane                   Bar           Airport Gate

#### Cluster 4 - Most common venues: Coffee shop, Restaurant, Café and Cocktail Bar

In [135]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
28  Downtown Toronto               3           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
28            Restaurant                  Café          Cocktail Bar   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
28    Italian Restaurant    Seafood Restaurant                 Hotel   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
28              Beer Bar        Breakfast Spot            Cheese Shop

#### Cluster 5 - Most common venues: Ice Cream Shop, Garden, Yoga Studio and Dog Run

In [136]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               4        Ice Cream Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22                Garden           Yoga Studio               Dog Run   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22   Filipino Restaurant  Fast Food Restaurant        Farmers Market   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22    Falafel Restaurant           Event Space   Ethiopian Restaurant